### PR TITLE
Signal GroupsV2 support in Auxin

### PR DESCRIPTION
**Pull request by NotGyro**
The feature/groups branch contains a proven-working implementation of message sending and message receiving for the Signal GroupsV2 protocol.

### CODE REVIEW

**Comment by itdaniher** [comment]
What's the biggest group we've tested in?

File: auxin/src/context.rs [hunk 1]
```diff
@@ -174,15 +185,8 @@ custom_error! { pub UnidentifiedAccessError
...
-	/// # Arguments
-	///
-	/// * `rng` - Mutable reference to a random number generator, which must be cryptographically-strong (i.e. implements the CryptoRng interface).
-	fn get_unidentified_access_unrestricted<R>(&mut self, rng: &mut R) -> crate::Result<Vec<u8>>
-	where
-		R: RngCore + CryptoRng,
-	{
-		let bytes: [u8; 16] = rng.gen();
+	fn get_unidentified_access_unrestricted(&mut self) -> crate::Result<Vec<u8>> {
+		let bytes = [0u8; 16];
```
Comment (DianaNites) [inline auxin/src/context.rs:189]: Why remove the random generation?

**Review by DianaNites — APPROVED** [review]
LGTM

File: auxin/src/context.rs (hunk 1 above)
Comment (NotGyro) [reply auxin/src/context.rs:189]: The way group message unidentified access keys work assumes "Unrestricted unidentified access" means all 0s. The group unidentified access key is derived by XORing every member's unidentified access key together, and if somebody's key is random that means it's not going to match what the server has.
See: [The relevant code path in Signal-Server
](https://github.com/signalapp/Signal-Server/blob/master/service/src/main/java/org/whispersystems/textsecuregcm/controllers/MessageController.java#L463)

**Pull request merged**